[GraphOptz] EliminateSliceConcat handles reshape

Squadrick opened the pull request:
Summary: If the slice dimension is not the same as concat dimension, it may be possible to replace the slices with a merged slice + reshape if the slice dimension is adjacent to the concat dimension.

Documentation: Comments added in code to explain new optimization.

Fixes #4628 

Test Plan: A new test case is added to check for this optimization: `EliminateSliceConcatWithReshapeTest`.

Squadrick commented:
cc @jfix71 for review

Squadrick commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
+                          std::vector<SliceNode *>>> consecutiveSlices;
     std::vector<SliceNode *> currConsecutiveSlices;
     SliceNode *lastSN = nullptr;
+    unsigned_t lastDim = -1;
```
Since it's unsigned, the default value will be `2**32-1`. I hope that's not a valid dimension I would have to consider.

Squadrick commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
     // Mapping from old Slices to new Slices where the range of each old Slice
     // is a subset of the corresponding new Slice
     std::unordered_map<SliceNode *, SliceNode *> oldToNewSlices;
-    for (const auto &slices : consecutiveSlices) {
+
+    // Mapping from new fused slices to reshape nodes that maybe be needed in
+    // case the fused slices need to be reshaped before concat. This happens
+    // when consecutive slices dimension differs from concat dimension by 1.
+    std::unordered_map<SliceNode *, ReshapeNode *> newSlicesToReshape;
```
Just thought of an alternate approach to create a single mapping as such:
```c++
std::unordered_map<SliceNode *, NodeValue *> oldToNewSlicesOrReshape;
```
and similarly, change the lookup logic. This would have a bit more casting but will use less memory (single hashmap vs. two). 

If warranted, I'll upload the new patch to have that change.

Squadrick replied:
Done. Much cleaner code. It uses `std::unordered_map<SliceNode *, Node *>`.

Squadrick commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
+      // concat dimension.
+      if (slicesDim == CN->getDim() + 1 ||
+          slicesDim == CN->getDim() - 1) {
+        auto outputDimVec = newSlice->getResult().dims().vec();
+        outputDimVec[CN->getDim()] *= outputDimVec[slicesDim];
+        outputDimVec[slicesDim] = 1;
```
Thought about this logic, and did a couple of experiment in Numpy, and it seems to hold. Still 100% sure, so some extra attention during the review may be warranted.

jfix71 replied:
Yeah, I think this is wrong. As an example, in your unit test, change the first few lines to the following, and it will fail:

```cpp
  auto *src =
      mod_.createPlaceholder(ElemKind::FloatTy, {8, 5, 3}, "src", false);
  auto *A = F_->createSlice("A", src, {0, 0, 0}, {2, 5, 3});
  auto *B = F_->createSlice("B", src, {2, 0, 0}, {4, 5, 3});
  auto *C = F_->createSlice("C", src, {4, 0, 0}, {6, 5, 3});
```

I think you want to use the original dimension of one of the slices.

Squadrick replied:
The optimization will only work if the size of the slice along the consecutive dimension is 1. This did bring up another problem, that I was not checking if this was the case. A check, along with a comment, has been added. So in your example, this optimization will do nothing.

jfix71 commented on tests/unittests/GraphOptzTest.cpp:
```diff
@@ -5451,6 +5451,31 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
   checkNumericalEquivalence();
 }
 
+TEST_F(GraphOptz, EliminateSliceConcatWithReshapeTest) {
```
Can you add an explanation of the test, e.g. perhaps with ASCII art? 🙂

jfix71 replied:
Can we also add another test a little more complex, e.g. with more inputs into the Concat other than just the three consecutive slices? Better yet, a Concat on dim1, with 3 consecutive slices on dim0 and another 3 consecutive slices on dim2.

jfix71 commented on tests/unittests/GraphOptzTest.cpp:
```diff
@@ -5451,6 +5451,31 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
+  optimizedF_ = optimizeFunction(
+      F_, {FunctionPassID::EliminateSliceConcat, getDCEPassConfig()});
+
+  // src, FUSED_SLICE[A, B, C], RESHAPE, CN.
+  // FUSED_SLICE, RESHAPE are inserted by optimization.
+  EXPECT_EQ(optimizedF_->getNodes().size(), 4);
```
Can you actually verify the structure of the graph instead of just putting what the expected nodes are after the opt? E.g. that all of the dimensions, concat axis (dim), etc. make sense.

jfix71 commented on tests/unittests/GraphOptzTest.cpp:
```diff
@@ -5451,6 +5451,31 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
+
+  bindings_.allocate(src)->getHandle<float>().randomize(-10.0, 10.0,
+                                                        mod_.getPRNG());
+  checkNumericalEquivalence();
```
You can specify the allowed error here -- let's make it 0.f since there's only data movement in this op.

jfix71 commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -2523,70 +2550,114 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
+        consecutiveSlices;
     std::vector<SliceNode *> currConsecutiveSlices;
     SliceNode *lastSN = nullptr;
+    unsigned_t lastDim = -1;
```
Do you encounter compilation issues if you use int instead of unsigned?

Squadrick replied:
Switched to `int`.

jfix71 commented on tests/unittests/GraphOptzTest.cpp:
```diff
@@ -5451,6 +5451,31 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
 }
 
+TEST_F(GraphOptz, EliminateSliceConcatWithReshapeTest) {
+  auto *src =
+      mod_.createPlaceholder(ElemKind::FloatTy, {4, 5, 3}, "src", false);
+  auto *A = F_->createSlice("A", src, {0, 0, 0}, {1, 5, 3});
+  auto *B = F_->createSlice("B", src, {1, 0, 0}, {2, 5, 3});
+  auto *C = F_->createSlice("C", src, {2, 0, 0}, {3, 5, 3});
```
As mentioned above:
```suggestion
  auto *src =
      mod_.createPlaceholder(ElemKind::FloatTy, {8, 5, 3}, "src", false);
  auto *A = F_->createSlice("A", src, {0, 0, 0}, {2, 5, 3});
  auto *B = F_->createSlice("B", src, {2, 0, 0}, {4, 5, 3});
  auto *C = F_->createSlice("C", src, {4, 0, 0}, {6, 5, 3});
```

Squadrick replied:
Explained in a previous comment. Only works if the slice size along the consecutive dimension is 1.

Squadrick commented:
@jfix71 Thanks for the review. Made the changes you requested. 

There is one significant change to the logic. If the slice consecutive dimension is after the concat dimension, there's a need for a transpose node: merged slice + transpose + reshape. Added the logic, and updated the comments. Could you take a look at that? 

The remaining work is around making the test better: ASIC art + check for structure. I'll push some commits soon for this.

Also, during graph optimization, what is the heuristic to optimize for? Is the sole focus to minimize the number of nodes, or should we take the type of node into consideration? Is it safe to assume that 4 slices should always be optimized to 1 slice + 1 transpose + 1 reshape?

Squadrick commented:
@jfix71 Did you get any time to take another look at this?

jfix71 commented:
Really sorry for the major delays here, I've been super swamped. I will definitely get to reviewing next week.

Squadrick commented:
@jfix71 No problem, take your time. Just didn't want the bot to auto-close for being stale for too long :).

jfix71 commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -2536,70 +2551,135 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
+        // 1) slices consecutive dimension is the same as concat dimension, or
+        // 2) slices consecutive dimension is adjacent to the concat dimension,
+        //    and the size of each slice along the consecutive dimension is 1.
+        //    NOTE: Checking the size of 0th slice is sufficient, as opposed
+        //    to checking every slice. If the slices can be concatenated,
+        //    each slice size must be equal.
```
Just to make this more clear, can we reword it to something like:

NOTE: Checking the slicesDim dimension of 0th slice is sufficient, as opposed to checking every slice. If the slices can be concatenated (and they're being concatenated along a different dimension), then each slicesDim dim must be equal.

jfix71 commented on tests/unittests/GraphOptzTest.cpp:
```diff
@@ -5525,6 +5525,38 @@ TEST_F(GraphOptz, EliminateSliceConcatTest) {
+
+  // src, FUSED_SLICE[A, B, C], RESHAPE, CN.
+  // FUSED_SLICE, RESHAPE are inserted by optimization.
+  EXPECT_EQ(optimizedF_->getNodes().size(), 9);
```
Can you do some actual testing of how many of each node is leftover? Using `countNodeKind()`, as seen in other tests.

Squadrick replied:
Done.

jfix71 reviewed (commented):
Hey @Squadrick, apologies again for all the delays. I think this is pretty close to landing -- one request would be to add in a test for any edge cases that the optimization should not apply for (e.g. the example I had specified in my comment, where the adjacent dims are not 1). We've had some opts in the past for slicing/concatenating where we ended up with numerical issues, and I'd like to make sure we have good coverage for cases both where the opt will and won't apply.

Squadrick reviewed (commented):
@jfix71 I've added another test: `EliminateSliceConcatWithReshapeTestNoChange` that ensures no optimization occurs when slice and concat dims are not adjacent. I've also made the changes you requested, including using `countNodeKind`.

jfix71 reviewed (approved):
LGTM, thanks for this contribution and your patience @Squadrick!
